internal/config/incremental_sync.go

```diff
@@ -0,0 +1,228 @@
+package config
+
+import (
+	"context"
+	"errors"
+	"github.com/icinga/icinga-go-library/database"
+	"github.com/icinga/icinga-go-library/types"
+	"github.com/jmoiron/sqlx"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"time"
+)
+
+// IncrementalConfigurable specifies Getter methods required for types supporting incremental configuration loading.
+type IncrementalConfigurable[T comparable] interface {
+	zapcore.ObjectMarshaler
+
+	// GetID returns the primary key value.
+	GetID() T
+
+	// GetChangedAt returns the changed_at value.
+	GetChangedAt() types.UnixMilli
+
+	// IsDeleted returns if this entry was marked as deleted.
+	IsDeleted() bool
+}
+
+// IncrementalConfigurableValidatable defines a validation method for new and updated elements within incrementalFetch.
+type IncrementalConfigurableValidatable interface {
+	// IncrementalValidate validates and optionally return an error, resulting in aborting this entry.
+	IncrementalValidate() error
+}
+
+// incrementalFetch queries all recently changed elements of BaseT and stores them in changeConfigSetField.
+//
+// The RuntimeConfig.configChangeTimestamps map contains the last known timestamp for each BaseT table. Only those
+// elements where the changed_at SQL column is greater than the stored timestamp will be fetched and stored in the
+// temporary RuntimeConfig.configChange ConfigSet. Later on, incrementalApplyPending merges it into the main ConfigSet.
+func incrementalFetch[
+	BaseT any,
+	CompT comparable,
+	T interface {
+		*BaseT
+		IncrementalConfigurable[CompT]
+	},
+](ctx context.Context, tx *sqlx.Tx, r *RuntimeConfig, changeConfigSetField *map[CompT]T) error {
+	startTime := time.Now()
+
+	var typePtr T
+
+	tableName := database.TableName(typePtr)
+	changedAt, hasChangedAt := r.configChangeTimestamps[tableName]
+
+	stmtLogger := r.logger.With(zap.String("table", tableName))
+
+	stmt := r.db.BuildSelectStmt(typePtr, typePtr)
+	stmtArgs := []any{}
+	if hasChangedAt {
+		stmtLogger = stmtLogger.With(zap.Time("changed_at", changedAt.Time()))
+		stmt += ` WHERE "changed_at" > ?`
+		stmtArgs = []any{changedAt}
+	}
```

Comment:
Yes, both for the initial loading as well as any subsequent loading if nothing was fetched during the first (or other previous) runs.

That's a good idea. In the current design, initially loaded deleted elements would result in a warning.
https://github.com/Icinga/icinga-notifications/blob/90f793f2bdf4fd9bb695b779be4799fcfbcbb38f/internal/config/incremental_sync.go#L161-L163
By doing so, this would be circumvented.